modules/packages/OrderedMap.chpl

```diff
@@ -0,0 +1,755 @@
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This module contains the implementation of the orderedMap type 
+  which is a container that stores key-value associations. 
+
+  orderedMaps are not parallel safe by default, but can be made parallel safe by
+  setting the param formal `parSafe` to true in any orderedMap constructor. When
+  constructed from another orderedMap, the new orderedMap will inherit 
+  the parallel safety mode of its originating orderedMap.
+
+  OrderedSet supports searching for a certain key, insertion and deletion in O(logN).
+*/
+module OrderedMap {
+  import ChapelLocks;
+  private use HaltWrappers;
+  private use OrderedSet;
+  private use IO;
+  public use Sort only defaultComparator;
+
+  // Lock code lifted from modules/standard/Lists.chpl.
+  pragma "no doc"
+  type _lockType = ChapelLocks.chpl_LocalSpinlock;
+
+  pragma "no doc"
+  class _LockWrapper {
+    var lock$ = new _lockType();
+
+    inline proc lock() {
+      lock$.lock();
+    }
+
+    inline proc unlock() {
+      lock$.unlock();
+    }
+  }
+
+  pragma "no doc"
+  proc _checkKeyType(type keyType) {
+    if isGenericType(keyType) {
+      compilerWarning("creating a orderedMap with key type " +
+                      keyType:string);
+      if isClassType(keyType) && !isGenericType(borrowed keyType) {
+        compilerWarning("which now means class type with generic management");
+      }
+      compilerError("orderedMap key type cannot currently be generic");
+    }
+  }
+
+  pragma "no doc"
+  proc _checkValType(type valType) {
+    if isGenericType(valType) {
+      compilerWarning("creating a orderedMap with value type " +
+                      valType:string);
+      if isClassType(valType) && !isGenericType(borrowed valType) {
+        compilerWarning("which now means class type with generic management");
+      }
+      compilerError("orderedMap value type cannot currently be generic");
+    }
+  }
+
+
+  /*
+    The shared nilable class helps us to find one element with the key
+    and without specifying the value
+    See `contains`
+  */
+  pragma "no doc"
+  class _valueWrapper {
+    var val;
+  }
+  /*
+    This doesn't indicate any comparing.
+    Just to make (keyType, shared _valueWrapper?) comparable.
+  */
+  proc <(a: shared _valueWrapper?, b: shared _valueWrapper?) {
+    return false;
+  }
+  proc >(a: shared _valueWrapper?, b: shared _valueWrapper?) {
+    return false;
+  }
```

Comment:
Are these operators needed?  The `_keyComparator` below should be used anywhere that tuple needs to be compared (I think).  And all the tests pass with them removed.

Reply:
If they are needed internally to the module, I think they should be declared private, and/or `no doc`.  And probably there should be a test that fails if they're not present.

Reply:
This could be taken care of after it's merged.